code/game/machinery/doors/firedoor.dm

```diff
@@ -188,30 +188,52 @@
 		for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
 			fire_panel.set_status()
 
+/obj/machinery/door/firedoor/proc/CalculateWatchedTurfs()
+	watched_turfs = list()
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(src),dir)
+		watched_turfs |= checked_turf
+		RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/check_atmos)
+	watched_turfs |= get_turf(src)
+
 /obj/machinery/door/firedoor/proc/check_atmos(datum/source)
+	SIGNAL_HANDLER
+	var/found_turf = FALSE
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+	
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
+	if(!watched_turfs)
+		watched_turfs = CalculateWatchedTurfs()
+
+	for(var/turf/checked_turf in watched_turfs)
+		if(source == checked_turf)
+			found_turf = TRUE
 
-	var/datum/gas_mixture/environment = my_turf.return_air()
 	var/result
 
-	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
-		result = FIRELOCK_ALARM_TYPE_HOT
-	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
-		result = FIRELOCK_ALARM_TYPE_COLD
-	if(!result)
-		return
+	for (var/turf/checked_turf in watched_turfs)
+		if(!checked_turf.density && found_turf)
+			var/datum/gas_mixture/environment = checked_turf.return_air()
+			if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
+				result = FIRELOCK_ALARM_TYPE_HOT
+			if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
+				result = FIRELOCK_ALARM_TYPE_COLD
```

Comment:
And this? It seems as if it'd be cleaner to maintain a list of not ok turfs, rather then iterating all our watched turfs each time the alarm sounds

Reply:
we only watch 4

Reply:
I think it'll make other problems easier too, but that's another discussion

Reply:
what do you mean exactly by a list of not okay turfs, everytime I experience an issue, add the not okay turf and check if it updates? wont I still need to check my surroundings